### ISSUE TITLE
FORWARD-TSN for partially sent message causes ABORT

### ISSUE THREAD

**Issue by boivie** [sctplab/usrsctp#592]
I'm not quite sure what is the issue here, but I want to understand why usrsctp sends an ABORT in this scenario:

```
A (running dcSCTP)

Z (running usrsctp)

# A starts to send a rather large message
A: DATA(TSN=1, B=1)
A: DATA(TSN=2)
A: DATA(TSN=3)

# A realizes that the message has now expired (limited lifetime), and will need to abort it by sending 
# a FORWARD-TSN, and continues sending the next message, a small non-fragmented one, in the 
# same SCTP packet
A: FORWARD-TSN(TSN=3) DATA(TSN=4, B=1, E=1)
Z: SACK(TSN=3)
Z: ABORT(Illegal message sequence, missing end for MID=3)
```

See attached PCAP files, but please note that you will need to click on the "Time" column in Wireshark to have the capture sorted by time, as the packet order in the PCAP file isn't correct (sorry about that). 

[usrsctp_forward_tsn.zip](https://github.com/sctplab/usrsctp/files/6541677/usrsctp_forward_tsn.zip)

Thanks in advance

**Comment by boivie** [comment]
I have been able to create a test case that reproduces the issue. I need to clean it up a bit, but in essence it does:

Create a link with 10% packet loss
for {
  sock_a.Send(10000 bytes, max_retransmissions=0) 
  sock_a.Send(10 bytes)
}

Attached logs show proper PCAP files (ordered items), for both sides, and full logs with verbose dcSCTP and verbose usrsctp logs.

[usrsctp-592.zip](https://github.com/sctplab/usrsctp/files/6545294/usrsctp-592.zip)

I will try to share the test case if you need it. Thanks!

**Comment by tuexen** [comment]
Let me try to reproduce this locally with `packetdrill` and report my findings later today.

**Comment by tuexen** [comment]
Here is a `packetdrill` script, which reproduces the behaviour you reported:
```
--tolerance_usecs=25000
 0.000 `sysctl -w net.inet.sctp.ecn_enable=0`
+0.000 `sysctl -w net.inet.sctp.pr_enable=1`
+0.000 `sysctl -w net.inet.sctp.asconf_enable=0`
+0.000 `sysctl -w net.inet.sctp.auth_enable=0`
+0.000 `sysctl -w net.inet.sctp.reconfig_enable=1`
+0.000 `sysctl -w net.inet.sctp.nrsack_enable=0`
+0.000 `sysctl -w net.inet.sctp.pktdrop_enable=0`
+0.000 socket(..., SOCK_STREAM, IPPROTO_SCTP) = 3
// Check the handshake with en empty(!) cookie
+0.000 bind(3, ..., ...) = 0
+0.000 listen(3, 1) = 0
+0.000 < sctp: INIT[flgs=0, tag=1, a_rwnd=5242880, os=65535, is=65535, tsn=0,
                    FORWARD_TSN_SUPPORTED[],
                    SUPPORTED_EXTENSIONS[types=[RECONFIG, FORWARD_TSN]]]
+0.000 > sctp: INIT_ACK[flgs=0, tag=2, a_rwnd=..., os=..., is=..., tsn=10,
                        FORWARD_TSN_SUPPORTED[],
                        SUPPORTED_EXTENSIONS[types=[FORWARD_TSN, RECONFIG]],
                        STATE_COOKIE[len=..., val=...]]
+0.010 < sctp: COOKIE_ECHO[flgs=0, len=..., val=...]
+0.000 > sctp: COOKIE_ACK[flgs=0]
+0.000 accept(3, ..., ...) = 4
+0.000 close(3) = 0
+0.000 < sctp: DATA[flgs=UBE, len=1016, tsn=0, sid=1, ssn=0, ppid=0]
+0.000 read(4, ..., 2000) = 1000
+0.000 > sctp: SACK[flgs=0, cum_tsn=0, a_rwnd=..., gaps=[], dups=[]]
+0.000 < sctp: DATA[flgs=UB, len=1176, tsn=1, sid=1, ssn=0, ppid=0]
+0.000 < sctp: DATA[flgs=U, len=1176, tsn=2, sid=1, ssn=0, ppid=0]
+0.000 > sctp: SACK[flgs=0, cum_tsn=2, a_rwnd=..., gaps=[], dups=[]]
+0.000 < sctp: DATA[flgs=U, len=1176, tsn=3, sid=1, ssn=0, ppid=0]
+0.000 < sctp: DATA[flgs=U, len=1176, tsn=4, sid=1, ssn=0, ppid=0]
+0.000 > sctp: SACK[flgs=0, cum_tsn=4, a_rwnd=..., gaps=[], dups=[]]
+0.000 < sctp: FORWARD_TSN[cum_tsn=4, ids=[]];
               DATA[flgs=UBE, len=20, tsn=5, sid=1, ssn=0, ppid=0]
+0.000 > sctp: ABORT[flgs=0, PROTOCOL_VIOLATION[info=...]]
+0.000 close(4) = 0
```
and here is a script which show the behaviour you expected:
```
--tolerance_usecs=25000
 0.000 `sysctl -w net.inet.sctp.ecn_enable=0`
+0.000 `sysctl -w net.inet.sctp.pr_enable=1`
+0.000 `sysctl -w net.inet.sctp.asconf_enable=0`
+0.000 `sysctl -w net.inet.sctp.auth_enable=0`
+0.000 `sysctl -w net.inet.sctp.reconfig_enable=1`
+0.000 `sysctl -w net.inet.sctp.nrsack_enable=0`
+0.000 `sysctl -w net.inet.sctp.pktdrop_enable=0`
+0.000 socket(..., SOCK_STREAM, IPPROTO_SCTP) = 3
// Check the handshake with en empty(!) cookie
+0.000 bind(3, ..., ...) = 0
+0.000 listen(3, 1) = 0
+0.000 < sctp: INIT[flgs=0, tag=1, a_rwnd=5242880, os=65535, is=65535, tsn=0,
                    FORWARD_TSN_SUPPORTED[],
                    SUPPORTED_EXTENSIONS[types=[RECONFIG, FORWARD_TSN]]]
+0.000 > sctp: INIT_ACK[flgs=0, tag=2, a_rwnd=..., os=..., is=..., tsn=10,
                        FORWARD_TSN_SUPPORTED[],
                        SUPPORTED_EXTENSIONS[types=[FORWARD_TSN, RECONFIG]],
                        STATE_COOKIE[len=..., val=...]]
+0.010 < sctp: COOKIE_ECHO[flgs=0, len=..., val=...]
+0.000 > sctp: COOKIE_ACK[flgs=0]
+0.000 accept(3, ..., ...) = 4
+0.000 close(3) = 0
+0.000 < sctp: DATA[flgs=UBE, len=1016, tsn=0, sid=1, ssn=0, ppid=0]
+0.000 read(4, ..., 2000) = 1000
+0.000 > sctp: SACK[flgs=0, cum_tsn=0, a_rwnd=..., gaps=[], dups=[]]
+0.000 < sctp: DATA[flgs=UB, len=1176, tsn=1, sid=1, ssn=0, ppid=0]
+0.000 < sctp: DATA[flgs=U, len=1176, tsn=2, sid=1, ssn=0, ppid=0]
+0.000 > sctp: SACK[flgs=0, cum_tsn=2, a_rwnd=..., gaps=[], dups=[]]
+0.000 < sctp: DATA[flgs=U, len=1176, tsn=3, sid=1, ssn=0, ppid=0]
+0.000 < sctp: DATA[flgs=U, len=1176, tsn=4, sid=1, ssn=0, ppid=0]
+0.000 > sctp: SACK[flgs=0, cum_tsn=4, a_rwnd=..., gaps=[], dups=[]]
+0.000 < sctp: FORWARD_TSN[cum_tsn=5, ids=[]];
               DATA[flgs=UBE, len=20, tsn=6, sid=1, ssn=0, ppid=0]
+0.000 read(4, ..., 2000) = 4
+0.200 > sctp: SACK[flgs=0, cum_tsn=6, a_rwnd=..., gaps=[], dups=[]]
// Tear down the association
+1.000 < sctp: SHUTDOWN[flgs=0, cum_tsn=9]
+0.000 > sctp: SHUTDOWN_ACK[flgs=0]
+0.000 < sctp: SHUTDOWN_COMPLETE[flgs=0]
+0.000 close(4) = 0
```
The point is that in the case you are reporting, your stack sends the first four fragments of an unordered user message using `tsn` of (let us say) 1, 2, 3, and 4. The all get processed by the SCTP stack and a corresponding `SACK` chunks is sent with `cum_tsn` of 4. Now your stack sends a `FORWARD_TSN` chunk with `cum_tsn` of 4. This means we don't have to do anything, since we have all fragments up to `tsn` 4. Now we get a `DATA` chunk with a complete unordered message using `tsn` 5. We analyse the situation, and detect that this is impossible, since you can't begin a new message without terminating the one before. That is why we send an `ABORT`. This behaviour is intended, and I think correct. One possible allowed TSN assignment is demonstrated in the second `packetdrill` script. Here a `DATA` chunk with `tsn` of 5 would have the terminating part of the user message and was abandoned by the stack.
Does this make things clearer?

**Comment by boivie** [comment]
Thank you very much Michael: The explanation makes much sense. I did not consider that, as dcSCTP would treat any FORWARD_TSN as deleting the received fragments for the currently received message which included a TSN as provided in the FORWARD-TSN (because that would be the intention by the sender, by sending a FORWARD-TSN). Re-reading RFC3758, that isn't covered and my interpretation can't really be defended reading the RFC.

Anyway: I did a local change where dcSCTP would add a (placeholder) abandoned fragment to its retransmission queue representing the end of the not-yet completely sent message which would become the cum_tsn in the sent FORWARD-TSN (just as your second packetdrill script), and then usrsctp accepts it.

This was a tricky one, and it would've taken me a long time to find this if you wouldn't have been here to help. Thanks a lot!
